src/identity/interaction/email-password/credentials/ClientCredentialsAdapterFactory.ts

```diff
@@ -0,0 +1,56 @@
+import type { AdapterPayload, Adapter } from 'oidc-provider';
+import type { KeyValueStorage } from '../../../../storage/keyvalue/KeyValueStorage';
+import type { AdapterFactory } from '../../../storage/AdapterFactory';
+import { PassthroughAdapterFactory, PassthroughAdapter } from '../../../storage/PassthroughAdapterFactory';
+
+export interface ClientCredentials {
+  secret: string;
+  webId: string;
+}
+
+/**
+ * A {@link PassthroughAdapter} that overrides the `find` function
+ * by checking if there are stored client credentials for the given ID
+ * if no payload is found in the source.
+ */
+export class ClientCredentialsAdapter extends PassthroughAdapter {
+  private readonly storage: KeyValueStorage<string, ClientCredentials>;
+
+  public constructor(name: string, source: Adapter, storage: KeyValueStorage<string, ClientCredentials>) {
+    super(name, source);
+    this.storage = storage;
+  }
+
+  public async find(id: string): Promise<AdapterPayload | void | undefined> {
+    let payload = await this.source.find(id);
+
+    if (!payload && this.name === 'Client') {
```

Comment:
So Client is a special name? Ought this to be documented somewhere?

Reply:
`Adapter` is an interface from the OIDC library and is documented here, including the possible names: https://github.com/panva/node-oidc-provider/blob/main/example/my_adapter.js

Reply:
Okay, so then the base typing is wrong apparently. Thanks!